src/lib.rs

```diff
@@ -371,6 +377,84 @@ pub trait ParallelDatabase: Database + Send {
     /// }
     /// ```
     fn snapshot(&self) -> Snapshot<Self>;
+
+    /// Returns a `Snapshot` which can be used to run a query concurrently
+    fn fork(&self, state: ForkState<Self>) -> Snapshot<Self>;
+
+    /// Returns a [`Forker`] object which can be used to fork new `DB` references that are able to
+    /// query the database concurrently. All queries run this way must complete before the
+    /// [`Forker`] object goes out of scope or its `Drop` impl will panic.
+    fn forker(&mut self) -> Forker<'_, Self> {
+        let runtime = self.salsa_runtime();
+        Forker {
+            state: ForkState(Arc::new(ForkStateInner {
+                parents: runtime
+                    .parent
+                    .iter()
+                    .flat_map(|state| state.0.parents.iter())
+                    .cloned()
+                    .chain(Some(runtime.id()))
+                    .collect(),
+                cycle: Default::default(),
+            })),
+            db: self,
+        }
+    }
+}
+
+/// Returned from calling [`ParallelDatabase::forker`]. Used to fork on a database so that
+/// multiple queries can run concurrently
+pub struct Forker<'a, DB>
+where
+    DB: Database,
+{
+    db: &'a mut DB,
+    state: ForkState<DB>,
+}
+
+///
+pub struct ForkState<DB: Database>(Arc<ForkStateInner<DB>>);
+
+struct ForkStateInner<DB: Database> {
+    parents: Vec<RuntimeId>,
+    cycle: Mutex<Vec<DB::DatabaseKey>>,
+}
+
+impl<DB: Database> Clone for ForkState<DB> {
+    fn clone(&self) -> Self {
+        ForkState(self.0.clone())
+    }
+}
+
+impl<DB> Drop for Forker<'_, DB>
+where
+    DB: Database,
+{
+    fn drop(&mut self) {
+        if !std::thread::panicking() {
```

Comment:
What about https://doc.rust-lang.org/std/mem/fn.forget.html

Reply:
Since there is no unsafe invariant to uphold forgetting the `Forker` just means that the query may end up in a bad state but there is no memory unsafety that is upheld by the drop.